### PR TITLE
Updated hard_tanh and added to layer_types

### PR DESCRIPTION
**Pull request by mrdaybird**
Related to #2777 
- [x] Updated hard_tanh.hpp 
- [x] Add tests for hard_tanh.hpp

### CODE REVIEW

**Comment by mrdaybird** [comment]
> Hey @mrdaybird , I would suggest you to refer this #2777 as you are missing some adaptions like copy and move function, openmp, virtual distructor, etc. :)

yeah you right, I kind of missed that. Thank you for pointing that out!

**Comment by mrdaybird** [comment]
@rcurtin Can you please review this PR?

File: src/mlpack/methods/ann/layer/hard_tanh_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+HardTanHType<MatType>& HardTanHType<MatType>::operator=(const HardTanHType& layer)
+{
+	if(&layer != this)
+	{
+		Layer<MatType>::operator=(layer);
+		maxValue = layer.maxValue;
+		minValue = layer.minValue;
+	}
+
+	return *this;
```
Comment (shubham1206agra) [inline src/mlpack/methods/ann/layer/hard_tanh_impl.hpp:60]: Please correct the spacing. See Design Guidelines for the same.

File: src/mlpack/methods/ann/layer/serialization.hpp [hunk 2]
```diff
@@ -69,6 +69,7 @@
     CEREAL_REGISTER_TYPE(mlpack::RBFType<__VA_ARGS__>); \
     CEREAL_REGISTER_TYPE(mlpack::SoftmaxType<__VA_ARGS__>); \
     CEREAL_REGISTER_TYPE(mlpack::SoftminType<__VA_ARGS__>); \
+	CEREAL_REGISTER_TYPE(mlpack::HardTanHType<__VA_ARGS__>); \
```
Comment (shubham1206agra) [inline src/mlpack/methods/ann/layer/serialization.hpp:72]: Here

File: src/mlpack/tests/ann/layer/hard_tanh.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+					   {1.0, 1.0, 0  , 0.0, 1.0},
+					   {1.0, 1.0, 1.0, 0.0, 1.0},
+					   {1.0, 0  , 0.0, 0.0, 0.0}};
+	gy.set_size(5,5);
+	gy.fill(1);
+	g.set_size(5,5);
+	//Test the Backward function
+	module.Backward(output, gy, g);
+	REQUIRE(arma::accu(g - delta) == Approx(0).epsilon(1e-4));
+}
```
Comment (shubham1206agra) [inline src/mlpack/tests/ann/layer/hard_tanh.cpp:55]: Here also

**Review by shubham1206agra — APPROVED** [review]
LGTM

**Comment by rcurtin** [comment]
Want to add a note to `HISTORY.md` too?

**Pull request merged**